rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,826 @@
+# Feature Name: `dev-tools-abstraction`
+
+## Summary
```

Comment:
Thank you for putting this together. I now understand where you all are coming from (for context: [my original comment pushing back against the "dev tool" concept](https://github.com/bevyengine/bevy/pull/12427#issuecomment-2005328288)).

I see the appeal of a common way to represent commands:

* Single searchable database of commands
* Common data model for command input

That being said, I think this is trying to be _too_ opinionated, and focusing on "dev tool" as a specific thing might be getting in our way:

1. Every "toolbox" will have a different way of interacting with these commands. Command serialization/deserialization is something the "toolbox" should be responsible for. We cannot just throw FromStr on the impl and let each dev tool trait impl choose what format to expect. We should rely on Reflect for this (with optional reflect-driven Serde support), similar to how we handle scene serialization. Leave things like the "command line format" to the "command line toolbox".
2. There is no reasonable way to solve the "Dev tool UI overlap" problem in a generic system. Each "toolbox" will define a different context for displaying and rendering "dev tools". The rendering/UX of a "dev tool" is _fully_ the responsibility of the toolbox, and most "interactive" functionality will be toolbox specific. A "dev-tool" that renders content inside of the Bevy Editor should define itself relative to the Bevy Editor's constraints (which are notably ... currently undefined). I do not think we should make solving this problem generically across toolboxes "future work". Instead it should be a non-goal.

I believe we should be fully focused on (1) a command-style abstraction and (2) queryable/categorized settings configured by commands. We should try very hard to build on what already exists when possible, and build reusable pieces when it isn't.

1. Reflect (with optional reflect-driven Serde) for all serialization
2. Use `Command` for "applying a command to a World" (I acknowledge that this is already part of the proposal)
3. Rely on existing registry tooling instead of building a new one. Some options:
    a. TypeRegistry enumeration of types registered with a specific reflected trait (ex: `DevCommand`)
    b. Make this Component / Query driven instead of Resource driven (ex: use a common DevCommand component)
4. Use Commands for configuring settings (including enabling / disabling features)

I think by shifting focus to these things, we sidestep the "dev tool is a nebulous concept" problem and focus on building the specific functionality we need. This also leaves the door open to other categories of commands / configuration.

For example (not a final design, just illustrating where my head is at):

```rust
// Reflect registry-driven commands

#[derive(Reflect, DevCommand, Debug)]
#[reflect(DevCommand)]
struct SetGold {
    amount: u64,
}

impl Command for SetGold {
    fn apply(self, world: &mut World){
        let mut current_gold = world.resource_mut::<Gold>();
        current_gold.0 = self.amount;
    }
}

app.register_type::<SetGold>()

// later, use the type registry to invoke / enumerate / serialize this command
type_registry.iter_with_data::<ReflectDevCommand>()
```

```rust
// Commands for enabling / disabling the settings of a dev tool

#[derive(Resource, Reflect, DevToolSettings, Debug)]
#[reflect(DevToolSettings)]
struct DevFlyCamera {
    enabled: bool,
    /// How fast the camera travels forwards, backwards, left, right, up and down, in world units.
    movement_speed: Option<f32>,
    /// How fast the camera turns, in radians per second.
    turn_speed: Option<f32>,
}

impl Toggleable for DevFlyCamera {
    fn is_enabled(&self) -> bool {
        self.enabled
    }
    
    fn set_enabled(&mut self, enabled: bool) {
        self.enabled = enabled;
    }
}

app.register_type::<DevFlyCamera>()
// Makes these Dev commands available (these could be put behind a helper function):
app.register_type::<Enable<DevFlyCamera>>()
app.register_type::<Disable<DevFlyCamera>>()
app.register_type::<Toggle<DevFlyCamera>>()
// this DevCommand writes a new value of DevFlyCamera on top of the existing value
app.register_type::<Configure<DevFlyCamera>>()

// later (ex: if you want to enumerate DevToolSettings types in a tool)
type_registry.iter_with_data::<DevToolSettings>()

/// Reusable Enable dev command for toggleable things
#[derive(DevCommand)]
pub struct Enable<T: Toggleable>();

impl<T: Toggleable> Command for Enable<T> {
    // TODO
}
```

Reply:
I can absolutely work with that! I'll do a clean up pass next week, reframing and simplifying to capture these ideas.

Reply:
Awesome!

Also note that `type_registry.iter_with_data` does not currently exist. You can do something similar with `.iter()` and filtering the results, but we could consider providing that ourselves (and _maybe_  building an index to accelerate, although building that index might be too detrimental to startup times and I'm not sure we need it.

Reply:
Me and @alice-i-cecile noticed during the re-designing of the abstraction to use `TypeRegistry` a few issues:

1. `TypeRegistry` was not built for this. The `TypeRegistry` was designed to store the registration of types, not tools or metadata of them. If we would use this, we are potentially allowing usage of `TypeRegistry` for storing **anything**.
2. Using a dedicated resource (e.g. DevToolsRegistry) is a standard design pattern in the engine. For example, let's take a look at [`DiagnosticStore`](https://github.com/bevyengine/bevy/blob/5357e15966b7936d1134df44b90f347b0f8e1edb/crates/bevy_diagnostic/src/diagnostic.rs#L295-L299) resource that stores diagnostic information. Similarly, a `DevToolsRegistry` could be used to store and manage tools and commands, while hiding a lot of *magic* that could be intimidating to users.
3. In the `DevToolsRegistry` we would store metadata about a tool or a command, which is accessible by providing its name. While this still *could* be possible with `TypeRegistry` it would be a misuse of the registry, and would create additional overhead.
4. `TypeRegistry` limits our flexibility and extensibility of the abstraction in the future.

The use of `Reflect` instead of `FromStr` is great though!

Reply:
> TypeRegistry was not built for this. The TypeRegistry was designed to store the registration of types, not tools or metadata of them. If we would use this, we are potentially allowing usage of TypeRegistry for storing anything.

Tools and metadata are represented as types, and _whether_ a type is a tool (or command) should be encoded as a Rust Trait. From my perspective these line up perfectly with the TypeRegistry's functionality and purpose.

> In the DevToolsRegistry we would store metadata about a tool or a command, which is accessible by providing its name. While this still could be possible with TypeRegistry it would be a misuse of the registry, and would create additional overhead.

From my perspective the "dev tool name" should be the type name (short name by default with disambiguation possible via the full type path). Again this feels like a great fit (and the right one). I consider the overhead of querying the reflected trait registration for a type to be acceptable for this use case (and _just barely_ more expensive than the `HashMap<String, Data>` that a custom registry would use).  

I think the biggest missing piece is how DevCommands correlate to dev tools / DevToolSettings (which we could just shorten to DevTool if we want). For example: the DevTool trait could have a function that provides this correlation.

> TypeRegistry limits our flexibility and extensibility of the abstraction in the future.

This was intentional. My goal here was to put bounds on this abstraction in a way that prevents it from becoming a new unbounded "framework". I'd like to reconcile the requirements of the "dev tool" space with the rest of Bevy.

Reply:
I agree here: I don't really see a use for Metadata anymore if we can use the type registry for many things
For example, let's say we have the `Toggleable` trait, and the `ReflectToggleable` type and the 3 new commands: `Toggle<T>`, `Enable<T>` and `Disable<T>`, and the CLI tool as an example of functionality here. 
We can have the CLI tool be just a `HashMap<ToolName, TypeId>` (tool name would probably be of type `String` here). We can just have the CLI tool use the name here and lookup in type registry for the `ReflectToggleable` data. For inserting new tools with CLI, we would use the same approach, but with `ReflectDeserialize` instead. 
And so, this would lessen the uses for metadata. Note that with this approach, the CLI tool for example would have to use serde with deserialize, as we don't really have a way to store the given function to deserialize the tool without some metadata. But with the main commands this RFC proposes: Enable, disable and toggle, this will go alright without metadata.
At the same time, I don't see a way to go without a `ToolBox` resource that will store at least the TypeId and the Tool name.

Reply:
> At the same time, I don't see a way to go without a ToolBox resource that will store at least the TypeIdand the Tool / command name.

That's why I believe we should have a registry. If we would ask every toolbox creator to "make your own storage for tools", we would end up with multiple resources that have the same purpose.

Reply:
> That's why I believe we should have a registry

Yeah, I believe we at can have a global resource to store what tools the app currently have, and rely more on the app registry for parse and creation of commands